presto-native-execution/presto_cpp/main/operators/tests/BroadcastTest.cpp

```diff
@@ -0,0 +1,177 @@
+ public:
+ protected:
+  void SetUp() override {
+    exec::test::OperatorTestBase::SetUp();
+    filesystems::registerLocalFileSystem();
+    exec::Operator::registerOperator(
+        std::make_unique<BroadcastWriteTranslator>());
+    exec::ExchangeSource::registerFactory(
+        BroadcastExchangeSource::createExchangeSource);
+  }
+
+  static std::string makeTaskId(const std::string& prefix, int num) {
+    auto url = fmt::format("batch/{}-{}", prefix, num);
+    return url;
+  }
+
+  std::shared_ptr<exec::Task> makeTask(
+      const std::string& taskId,
+      core::PlanNodePtr planNode,
+      int destination) {
+    auto queryCtx = std::make_shared<core::QueryCtx>(executor_.get());
+    core::PlanFragment planFragment{planNode};
+    return exec::Task::create(
+        taskId, std::move(planFragment), destination, std::move(queryCtx));
+  }
+
+  auto addBroadcastWriteNode(const std::string& basePath) {
+    return [&basePath](
+               core::PlanNodeId nodeId,
+               core::PlanNodePtr source) -> core::PlanNodePtr {
+      return std::make_shared<BroadcastWriteNode>(
+          nodeId, basePath, std::move(source));
+    };
+  }
+
+  auto addBroadcastReadNode(const velox::RowTypePtr& outputType) {
+    return [&outputType](
+               core::PlanNodeId nodeId,
+               core::PlanNodePtr /* source */) -> core::PlanNodePtr {
+      return std::make_shared<core::ExchangeNode>(nodeId, outputType);
+    };
+  }
+
+  void runBroadcastTest(const std::vector<RowVectorPtr>& data) {
+    exec::Operator::registerOperator(
+        std::make_unique<BroadcastWriteTranslator>());
+
+    // Flatten the inputs to avoid issues assertEqualResults referred here:
+    // https://github.com/facebookincubator/velox/issues/2859
+    auto dataType = asRowType(data[0]->type());
+    std::vector<RowVectorPtr> flattenInputs;
+    for (auto& input : data) {
+      flattenInputs.push_back(vectorMaker_.flatten<RowVector>(input));
+    }
+
+    auto writerTaskId = makeTaskId("leaf", 0);
+    auto basePath = fmt::format("/tmp/{}", writerTaskId);
+    auto writerPlan =
+        exec::test::PlanBuilder()
+            .values(flattenInputs, true)
+            .addNode(addBroadcastWriteNode(basePath))
+            // Todo - Fix why we are not getting pages from below node
+            //.partitionedOutputBroadcast()
+            .planNode();
+
+    exec::test::CursorParameters params;
+    params.planNode = writerPlan;
+    auto [taskCursor, results] = readCursor(params, [](auto /*task*/) {});
+
+    // Expect one file for each request
+    ASSERT_EQ(results.size(), 1);
+
+    // Validate file path prefix is consistent
+    auto broadcastFilePath = results.back()->childAt(0)->toString(0);
+    ASSERT_TRUE(boost::starts_with(broadcastFilePath, basePath));
+
+    // Create plan for read node using file path
+    auto readerPlan = exec::test::PlanBuilder()
+                          .addNode(addBroadcastReadNode(dataType))
```

Comment:
Any particular reason not to use PlanBuilder::exchange method?

```
.exchange(dataType)
```

Reply:
It works either way. Cleaned it up to use exchange